### PR TITLE
Support rtx for publisher

### PR DESCRIPTION
**Pull request by cnderrauber**

### CODE REVIEW

File: pkg/rtc/transport.go [hunk 1]
```diff
@@ -326,6 +331,37 @@ func newPeerConnection(params TransportParams, onBandwidthEstimator func(estimat
...
+		if twccExtID != 0 {
+			if buffer := params.Config.BufferFactory.GetBuffer(info.SSRC); buffer != nil {
+				params.Logger.Debugw("set rtx twcc and ext id", "ssrc", info.SSRC, "twccExtID", twccExtID)
+				buffer.SetTWCCAndExtID(params.Twcc, uint8(twccExtID))
+			} else {
+				params.Logger.Errorw("failed to get buffer for rtx stream", nil, "ssrc", info.SSRC)
+			}
+		}
+	}
+	// put rtx inteceptor behind unhandle simulcast interceptor so it can get the correct mid & rid
```
Comment (boks1971) [inline pkg/rtc/transport.go:360]: :nit: `interceptor` spelling

File: pkg/rtc/transport.go [hunk 2]
```diff
@@ -326,6 +331,37 @@ func newPeerConnection(params TransportParams, onBandwidthEstimator func(estimat
 			ir.Add(f)
 		}
 	}
+
+	setTWCCForVideo := func(info *interceptor.StreamInfo) {
```
Comment (boks1971) [inline pkg/rtc/transport.go:335]: :nit: just call this `setTWCC`. Although it is used only for `video`, does not need to be only video.

File: pkg/sfu/interceptor/rtx.go [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+
+type RTXInfoExtractor struct {
+	interceptor.NoOp
+
+	factory *RTXInfoExtractorFactory
+	logger  logger.Logger
+}
+
+func (u *RTXInfoExtractor) BindRemoteStream(info *interceptor.StreamInfo, reader interceptor.RTPReader) interceptor.RTPReader {
+	if !strings.HasPrefix(info.MimeType, "video") {
```
Comment (boks1971) [inline pkg/sfu/interceptor/rtx.go:101]: should we restrict this to `video`? Okay for now, but if possible, would be good to just keep it generic as audio NACKs are potentially use the RTX stream too.

File: pkg/sfu/interceptor/rtx.go [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+	"github.com/pion/webrtc/v3"
+
+	"github.com/livekit/livekit-server/pkg/sfu/utils"
+	"github.com/livekit/protocol/logger"
+)
+
+const (
+	SDESRepairRTPStreamIDURI = "urn:ietf:params:rtp-hdrext:sdes:repaired-rtp-stream-id"
+
+	rtxProbeCount = 10
```
Comment (boks1971) [inline pkg/sfu/interceptor/rtx.go:18]: is 10 probes enough? Looks like probe tries is decremented if either mid && rid || rsid is not empty or if not a padding packet. Seems like primary stream could exhaust it?

**Review by boks1971 — APPROVED** [review]
Great work @cnderrauber . Some minor comments.

File: pkg/sfu/interceptor/rtx.go (hunk 4 above)
Comment (cnderrauber) [reply pkg/sfu/interceptor/rtx.go:18]: The primary stream and rtx stream are two streams and will not share probe counts, so it is to probe rtx stream after exclude padding packet.

File: pkg/sfu/interceptor/rtx.go (hunk 3 above)
Comment (cnderrauber) [reply pkg/sfu/interceptor/rtx.go:101]: No need to restrict it here, move to `setTWCC` callback

File: pkg/rtc/transport.go (hunk 2 above)
Comment (cnderrauber) [reply pkg/rtc/transport.go:335]: Chrome's offer contains audio twcc fb and extension too and currently implementation is only enable twcc for video. I think enabling audio twcc is unrelate to RTX support, so keeps the behavior in this PR.

**Comment by jadbox** [comment]
what does this do? How do I use it?

**Pull request merged**